fdbclient/include/fdbclient/BlobCipher.h

```diff
@@ -181,6 +191,212 @@ struct hash<BlobCipherDetails> {
+	                 const uint8_t* ivBuf,
+	                 const int ivLen,
+	                 Arena& arena)
+	  : cipherTextDetails(textDetails), cipherHeaderDetails(headerDetails) {
+		ASSERT_EQ(ivLen, AES_256_IV_LENGTH);
+		memcpy(&iv[0], ivBuf, ivLen);
+		memset(&authToken[0], 0, AuthTokenSize);
+	}
+
+	bool operator==(const AesCtrWithAuthV1<AuthTokenSize>& o) const {
+		return cipherHeaderDetails == o.cipherHeaderDetails && cipherTextDetails == o.cipherTextDetails &&
+		       memcmp(&iv[0], &o.iv[0], AES_256_IV_LENGTH) == 0 &&
+		       memcmp(&authToken[0], &o.authToken[0], AuthTokenSize) == 0;
+	}
+
+	static StringRef toStringRef(const AesCtrWithAuthV1<AuthTokenSize>& algoHeader, Arena& arena) {
+		BinaryWriter wr(AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+		wr.serializeBytes(&algoHeader, sizeof(AesCtrWithAuthV1<AuthTokenSize>));
+		return wr.toValue(arena);
+	}
+
+	static AesCtrWithAuthV1<AuthTokenSize> fromStringRef(const StringRef& algoHeaderRef, Arena& arena) {
+		AesCtrWithAuthV1<AuthTokenSize> algoHeader;
+		BinaryReader rd(
+		    algoHeaderRef.begin(), algoHeaderRef.size(), AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+		rd >> algoHeader;
+		return algoHeader;
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		ar.serializeBytes(this, sizeof(AesCtrWithAuthV1<AuthTokenSize>));
+	}
+};
+
+struct AesCtrNoAuthV1 {
+	// Serializable fields
+
+	// Text cipher encryption information
+	BlobCipherDetails cipherTextDetails;
+	// Initialization vector
+	uint8_t iv[AES_256_IV_LENGTH];
+
+	AesCtrNoAuthV1() {}
+	AesCtrNoAuthV1(const BlobCipherDetails& textDetails, const uint8_t* ivBuf, const int ivLen, Arena& arena)
+	  : cipherTextDetails(textDetails) {
+		ASSERT_EQ(ivLen, AES_256_IV_LENGTH);
+		memcpy(&iv[0], ivBuf, ivLen);
+	}
+
+	bool operator==(const AesCtrNoAuthV1& o) const {
+		return cipherTextDetails == o.cipherTextDetails && memcmp(&iv[0], &o.iv[0], AES_256_IV_LENGTH) == 0;
+	}
+
+	static StringRef toStringRef(const AesCtrNoAuthV1& noAuth, Arena& arena) {
+		BinaryWriter wr(AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+		wr.serializeBytes(&noAuth, sizeof(AesCtrNoAuthV1));
+		return wr.toValue(arena);
+	}
+
+	static AesCtrNoAuthV1 fromStringRef(const StringRef& headerRef, Arena& arena) {
+		AesCtrNoAuthV1 noAuth;
+		BinaryReader rd(headerRef.begin(), headerRef.size(), AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+		rd >> noAuth;
+		return noAuth;
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		ar.serializeBytes(this, sizeof(AesCtrNoAuthV1));
+	}
+};
+
+struct BlobCipherEncryptHeaderRef {
+	constexpr static FileIdentifier file_identifier = 1288927;
+
+	// Serializable fields
+
+	// HeaderFlags version tracker
+	uint16_t flagsVersion;
+	// Encryption algorithm header version tracker
+	uint16_t algoHeaderVersion;
+	// Serialized HeaderFlags
+	StringRef flagsRef;
+	// Serialized encryption algorithm tracker
+	StringRef algoHeaderRef;
```

Comment:
The following will do basically the same thing but it do without custom logic to serialize structs as StringRef and I believe is closer to how binary serializer are to be used. And we can further make `std::variant` a natively supported type of binary serializer, despite that would add a 1-byte size for serializing it.
```
struct BlobCipherEncryptHeaderRef {
    uint8_t headerVersion = ENCRYPT_HEADER_FLAGS_VERSION;
    uint8_t encryptMode;
    uint8_t authTokenMode;
    uint8_t authTokenAlgo;
    std::variant<AesCtrWithAuthRef, AesCtrNoAuthRef, ...> algoHeader;

    template <class Ar>
    void serializer(Ar& ar) {
        serializer(ar, headerVersion);
        if (headerVersion == 1) {
            serializer(ar, encryptMode, authTokenMode, authTokenAlgo);
        } else {
            ASSERT(false);
        }
        if (isAesCtr(...) {
            if (withAuth(...)) {
                AesCtrWithAuthRef aesWithAuth;
                serializer(ar, aesWithAuth);
                algoHeader = aesWithAuth;
            } else {
                ... // aesCtrNoAuth
            }
        } else {
            ... // other algo
        }
    }
};

struct AesCtrNoAuthRef {
    uint8_t version = ENCRYPT_HEADER_AES_CTR_NO_AUTH_VERSION;
    BlobCipherDetails details;
    StringRef iv;

    template <class Ar>
    void serialize(Ar& ar) {
        serializer(ar, version);
        if (version == 1) {
           serializer(details, iv);
           ASSERT_EQ(AES_256_IV_LENGTH, iv.size());
        } else {
            ASSERT(false);
        }
    }
};

```

Reply:
Sounds good. I like this idea of enriching the (de)serialize method to figure out what needs to be written to the disk as well as memory-representation needs to be when reading out from the disk.. Will give it a try..

Btw, AesCtr with/without Auth, don't have any StringRef, hence, only top-level encryption header needs to make the right decision. Also, if it works, it avoids ObjectSerialization even for the top level header.. 

Thanks

Reply:
btw I made `iv` (same for `authToken`) a StringRef above, but after a second thought I think keeping them as `uint8_t[]`, as this PR is doing, is probably better, since they are of fixed size.

Reply:
Updated the code to do the following:
1. Leverage `std::variant` in-memory representation Flags and AlgoHeader
2. Binary (de)serialization is now used for all BlobEncryptCipher header structs (parent and contained structs)

As expected the header sizes are looking even more better 

------------------------------------------------------------------------------------------------
|         S.No                |   ObjFlags       |    BinaryFlags        |    ObjectAlgo     |       BinaryAlgo     |   TotalObject   |    TotalBinary   |
| :------------------ | :-----------    | :----------            | :---------   | :-----------   | :--------  | :-------- |
|   AesCtrNoAuth       |     40               |       3                      |       104       |       40            |     208      |    55      |
|   AesCtrHmacSha   |     40               |       3                      |       184       |       96           |     288      |    111      |
|   AesCtrAesCmac   |     40               |       3                       |       168       |       80           |     272      |    95       |
------------------------------------------------------------------------------------------------